EIPS/eip-3540.md

```diff
@@ -47,44 +47,40 @@ In order to guarantee that every EOF-formatted contract in the state is valid, w
-- otherwise code is considered *legacy code* and the following rules do not apply to it.
+We introduce *code validation* for new contract creation, which EOF formatted code will undergo. To achieve this, we define a format called EVM Object Format (EOF), containing a version indicator, and a ruleset of validity tied to a given version.
 
-For a create transaction, if *initcode* or *code* is invalid, the contract creation results in an exceptional abort. Such a transaction is valid and may be included in a block. Therefore, the transaction sender's nonce is increased.
+Legacy code is not affected by EOF code validation.
 
-For the `CREATE` and `CREATE2` instructions, if *initcode* or *code* is invalid, instructions' execution ends with the result `0` pushed on stack. The *initcode* validation happens just before its execution and validation failure is observable as if execution results in an exceptional abort. I.e. in case *initcode* or returned *code* is invalid the caller's nonce remains increased and all creation gas is deducted.
+Code validation is performed during `CREATE4` instruction, and is elaborated on in [EIP-3670](./eip-3670.md) and an upcomming contract creation EIP-???? (TBD, refer to the [Megaspec doc](https://github.com/ipsilon/eof/blob/main/spec/eof.md) until then)
+The EOF format itself and its formal validation are described in the following sections.
 
 ### Container specification
 
 EOF container is a binary format with the capability of providing the EOF version number and a list of EOF sections.
 
-The container starts with the EOF header:
+The container starts with the EOF prefix:
 
 | description | length   | value      |                    |
 |-------------|----------|------------|--------------------|
 | magic       | 2-bytes  | 0xEF00     |                    |
 | version     | 1-byte   | 0x01–0xFF  | EOF version number |
 
-The EOF header is followed by at least one section header. Each section header contains two fields, `section_kind` and either `section_size` or `section_size_list`, depending on the kind. `section_size_list` is a list of size values when multiple sections of this kind are allowed.
+The EOF prefix is followed by at least one section header. Each section header contains two fields, `section_kind` and bytes describing the section of structure defined specifically for each kind.
 
-| description       | length  | value         |                   |
-|-------------------|---------|---------------|-------------------|
-| section_kind      | 1-byte  | 0x01–0xFF     | `uint8`           |
-| section_size      | 2-bytes | 0x0000–0xFFFF | `uint16`          |
-| section_size_list | dynamic | n/a           | `uint16, uint16+` |
+| description         | length  | value         |                   |
+|---------------------|---------|---------------|-------------------|
+| section_kind        | 1-byte  | 0x01–0xFF     | `uint8`           |
+| section description | dynamic | n/a           | `uint8+`          |
```

Comment:
Not sure why this part was changed, I think previous general description still applies?

Reply:
It doesn't - for code and containers sections it is not a 2-byte size anymore.

Reply:
`section_size` applied only to single-size sections (type and data), and `section_size_list` applied to code sections and now should also apply to container sections. The paragraph above described this well:

> The EOF header is followed by at least one section header. Each section header contains two fields, `section_kind` and either `section_size` or `section_size_list`, depending on the kind. `section_size_list` is a list of size values when multiple sections of this kind are allowed.

Reply:
Oh, so `_list` here included the number of items as the first `uint16` :facepalm: . I somehow misread this and the "or", and it seemed completely out-of-date

Nevertheless. I thought we wanted to make the general EOF header format more liberally defined, per the rationale included below:

> - Whether or not to have more structure to the container header for all EOF versions to follow. In order to allow future formats optimized for chunking and merkleization (verkleization) it was decided to keep it generic and specify the structure only for a specific EOF version.

This would invite putting offsets in, rather than sizes.

Should we revert this and require a specific structure for all versions of EOF?

Reply:
Agreed on the call to revert, https://github.com/ethereum/EIPs/pull/8152/commits/7e4c93bfc507dd84a59cb6dfe02f8a6711e84258 (I also took the liberty to be more specific about how the list is encoded)